guardian/mobile-n10n
# Sender lambda custom execution context

## What does this change?

As discussed in https://github.com/guardian/mobile-n10n/pull/727 there can be performance improvements by increasing the thread pool size.

We did some testing in code to evaluate the change:
- we defined new SSM variables for the thread pool size, allowing us to easily increase/decrease the value (without requiring code changes)
- we saw no errors or out memory issues when testing in code and identified the key metrics that allow us to understand if the change is successful

Before this change is merged we need to define new parameters in SSM:
- /notifications/PROD/workers/ios/apns.threadPoolSize
- /notifications/PROD/workers/ios-edition/apns.threadPoolSize
- /notifications/PROD/workers/android/fcm.threadPoolSize
- /notifications/PROD/workers/android-edition/fcm.threadPoolSize
- /notifications/PROD/workers/android-beta/fcm.threadPoolSize

It's suggested to initially set all the thread pools to a size of 10.

Once deployed we will monitor the following metrics:
- memory usage of the lambdas
- alerts or errors from the lambdas
- processing time / 90in2 statistic

It's suggested that we increase only the thread pool size for the `ios` and `android` workers. It's suggested to gradually increase the thread pool size (over a number of days) and monitor the results.

## How to test

Once deployed ensure lambdas correctly initialise (including all ssm variables being pulled from systems manager).

Evaluate the metrics mentioned above and check for any related alerts.

## Have we considered potential risks?

There can be risks by increasing the thread pool size, e.g. out of memory issues. We intend to gradually increase the thread pool size. We will only increase if we see performance improvements over a number of days. We can easily change the thread pool size by changing the configuration in SSM (no code change required).
